### DISCUSSION TITLE
Avoiding resampling where possible

### DISCUSSION THREAD

**Discussion by IcyAlmond** [wwmm/easyeffects#1517] [Q&A] [answered]
Does easyeffects resample if the application and pipewire have mismatching sample rates?

Is there a way to make a best effort to avoid resampling or is this an inherent flaw of pipewire?

**Comment by wwmm** [comment, marked as answer]
As far as I remember there is only one situation where EasyEffects has to do resampling. It is when PipeWire is not operating at 48 kHz and the `noise reduction` plugin is in the pipeline. This plugin works only with 48 kHz.

Nowadays PipeWire does dynamic sampling rate switching when possible. And it adjusts the rate used in our filters automatically in these cases. So a "mismatch" is not going to happen.

**Reply by IcyAlmond** [reply]
If an application is at for eg 41.1kHz but the pipewire server is at 48kHz, what happens in the case the noise reduction plugin is being used and the case that plugin doesnt exist

**Reply by wwmm** [reply]
`41.1 kHz`? Or did you mean `44.1 kHz`? I doubt there are sound cards out there that support `41.1 kHz` natively. In this case the sound server will have to do resampling because the hardware does not support the rate.

But if you meant `44.1 kHz` something else may happen. If this is the only player running PipeWire will try to set the card to 44.1 kHz in order to avoid resampling. In the past both PipeWire and Pulseaudio were only able to do that if the card had a chance to enter the idle state before the player with the desired rate started playing. But I think nowadays this limitation may not exist anymore. I am not sure.

**Comment by wwmm** [comment]
That being said there are situations where resampling is necessary. For example when you have 2 players with different rates playing to the same device. One of them will have to be resampled. It is not an audio stack limitation. It is how the math of audio processing works.
